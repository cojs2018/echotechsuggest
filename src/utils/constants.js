export const bookmarksAPI = 'https://5zcbdoz0h3.execute-api.eu-west-2.amazonaws.com/dev';
export const tagsAPI = 'https://141b51xkk2.execute-api.eu-west-2.amazonaws.com/dev';