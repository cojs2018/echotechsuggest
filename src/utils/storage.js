import { writeFile } from 'fs';

export async function createBookmark(bookmarkUrl) {
    try {
        let bookmarkUrlList = require('../../bookmarkUrls.json');

        let bookmarkLocation = bookmarkUrl;
        if(!bookmarkLocation.includes('http')) {
            bookmarkLocation = `http://${bookmarkLocation}`;
        }

        const newUrlData = new URL(bookmarkLocation);

        const bookmarkJSON = {
            path: bookmarkUrl,
            url: newUrlData.toJSON(),
            createdAt: Date.now()
        }
        const articleName = `${newUrlData.pathname} - ${newUrlData.hostname}`;

        bookmarkUrlList[articleName] = bookmarkJSON;

        writeFile(
            '../../bookmarkUrls.json',
            JSON.stringify(bookmarkJSON, ' ', 4)
        )

        return {
            status: 200,
            message: 'Bookmark created successfully!'
        }
    }
    catch (error) {
        throw new Error(JSON.stringify({ 
            status: 500,
            message: 'Internal service error: Could not create bookmark!' 
        }));
    }
}

export async function listBookmarks() {
    const bookmarkUrlList = require('../../bookmarkUrls.json');
    return bookmarkUrlList;
}

export async function deleteBookmark(bookmarkName) {
    try {
        let bookmarkUrlList = require('../../bookmarkUrls.json');
        
        if(!bookmarkUrlList[bookmarkName]) {
            return {
                status: 200,
            }
        }

        delete bookmarkUrlList[bookmarkName]

        writeFile(
            '../../bookmarkUrls.json',
            JSON.stringify(bookmarkJSON, ' ', 4)
        )

        return {
            status: 200,
            message: 'Bookmark deleted successfully!'
        }
    }
    catch (error) {
        throw new Error(JSON.stringify({ 
            status: 500,
            message: 'Internal service error: Could not delete bookmark!' 
        }));
    }
}