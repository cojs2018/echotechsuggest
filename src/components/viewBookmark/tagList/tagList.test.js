jest.mock('../../../utils/storage')

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import TagList from './tagList';
import { listTags, updateBookmark } from '../../../utils/storage';

listTags.mockReturnValue(new Promise((resolve, reject) => {
    resolve([
        {
            "tagId": "0d411a0b-da77-4f30-a6dd-c2deda6618c0",
            "value": "Artificial Intelligence",
            "featured": 4
        },
        {
            "tagId": "b659f09b-9cc7-442b-aa80-bede44772ac4",
            "value": "Machine Learning",
            "featured": 1
        },
    ])
}))

describe('TagList', () => {

    const props = {
        bookmarkId: '047ec796-2eca-417e-865a-e2e187b97689',
        currentTags: [
            {
                "tagId": "94972406-7158-4a27-8a44-52a33b854a74",
                "value": "Artificial Precipitation",
                "featured": 1
            },
            {
                "tagId": "18a2d966-fe32-4d05-9543-fcee703f718c",
                "value": "Energy",
                "featured": 1
            },
            {
                "tagId": "52d91ec3-a18b-4864-8438-56cfd759528c",
                "value": "Quantum Computing",
                "featured": 2
            },
            {
                "tagId": "8b48abe0-98df-4c33-99d4-543ff3f4bf54",
                "value": "Biotechnology",
                "featured": 1
            }
        ], 
        setAlertMessage: jest.fn(async () => {}),
        setVisible: jest.fn(async () => {}),
    }

    it('renders correctly', () => {
        const tree = renderer
            .create(<TagList {...props} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    })
});

test('Edit', async () => {

    const props = {
        bookmarkId: '047ec796-2eca-417e-865a-e2e187b97689',
        currentTags: [], 
        setAlertMessage: jest.fn(async () => {}),
        setVisible: jest.fn(async () => {}),
    };

    let tagSelection = [];
    let newTags = [];
    let edit = false;
    let dropDown = false;

    const setTagSelection = jest.fn(async (newTagsSelection) => tagSelection = newTagsSelection);
    const setNewTags = jest.fn(async (newTagSet) => newTags = newTagSet);
    const setEdit = jest.fn(async (bool) => edit = bool);
    const setDropDown = jest.fn(async (bool) => dropDown = bool);

    React.useEffect = jest.fn()
        .mockImplementation((effect, deps = null) => effect());
    
    React.useState = jest.fn()
        .mockReturnValueOnce([tagSelection, setTagSelection])
        .mockReturnValueOnce([newTags, setNewTags])
        .mockReturnValueOnce([edit, setEdit])
        .mockReturnValueOnce([dropDown, setDropDown]);

    const wrapper = mount(<TagList {...props} />);

    const editIcon = wrapper
        .find('IconButton')
        .first();

    editIcon.props().onPress();
    wrapper.update();

    await new Promise(resolve => setImmediate(resolve));
    expect(edit).toBeTruthy();
});

test('Cancel', async () => {

    let alertMessage = {};
    let visible = '';

    const props = {
        bookmarkId: '047ec796-2eca-417e-865a-e2e187b97689',
        currentTags: [], 
        setAlertMessage: jest.fn(async (newMessage) => alertMessage = newMessage),
        setVisible: jest.fn(async (bool) => visible = bool),
    };

    let tagSelection = [];
    let newTags = [];
    let edit = true;
    let dropDown = false;

    const setTagSelection = jest.fn(async (newTagsSelection) => tagSelection = newTagsSelection);
    const setNewTags = jest.fn(async (newTagSet) => newTags = newTagSet);
    const setEdit = jest.fn(async (bool) => edit = bool);
    const setDropDown = jest.fn(async (bool) => dropDown = bool);

    React.useEffect = jest.fn()
        .mockImplementation((effect, deps = null) => effect());
    
    React.useState = jest.fn()
        .mockReturnValueOnce([tagSelection, setTagSelection])
        .mockReturnValueOnce([newTags, setNewTags])
        .mockReturnValueOnce([edit, setEdit])
        .mockReturnValueOnce([dropDown, setDropDown]);

    const wrapper = mount(<TagList {...props} />);

    const cancelIcon = wrapper
        .find('IconButton')
        .filterWhere(iconButton => iconButton.props().icon === 'cancel')
        .first();

    cancelIcon.props().onPress();
    wrapper.update();

    await new Promise(resolve => setImmediate(resolve));
    expect(edit).toBeFalsy();
});

test('Remove Tag', async () => {

    let alertMessage = {};
    let visible = '';

    const props = {
        bookmarkId: '047ec796-2eca-417e-865a-e2e187b97689',
        currentTags: [], 
        setAlertMessage: jest.fn(async (newMessage) => alertMessage = newMessage),
        setVisible: jest.fn(async (bool) => visible = bool),
    };

    let tagSelection = [];
    let newTags = [
        {
            "tagId": "52d91ec3-a18b-4864-8438-56cfd759528c",
            "value": "Quantum Computing",
            "featured": 2
        }
    ];
    let edit = true;
    let dropDown = false;

    const setTagSelection = jest.fn(async (newTagsSelection) => tagSelection = newTagsSelection);
    const setNewTags = jest.fn(async (newTagSet) => newTags = newTagSet);
    const setEdit = jest.fn(async (bool) => edit = bool);
    const setDropDown = jest.fn(async (bool) => dropDown = bool);

    React.useEffect = jest.fn()
        .mockImplementation((effect, deps = null) => effect());
    
    React.useState = jest.fn()
        .mockReturnValueOnce([tagSelection, setTagSelection])
        .mockReturnValueOnce([newTags, setNewTags])
        .mockReturnValueOnce([edit, setEdit])
        .mockReturnValueOnce([dropDown, setDropDown]);

    const wrapper = mount(<TagList {...props} />);

    const tagChip = wrapper
        .find('Chip')
        .first();

    tagChip.props().onClose();
    wrapper.update();

    await new Promise(resolve => setImmediate(resolve));
    expect(newTags.length).toStrictEqual(0);
});

test('New Tag', async () => {

    let alertMessage = {};
    let visible = '';

    const props = {
        bookmarkId: '047ec796-2eca-417e-865a-e2e187b97689',
        currentTags: [], 
        setAlertMessage: jest.fn(async (newMessage) => alertMessage = newMessage),
        setVisible: jest.fn(async (bool) => visible = bool),
    };

    let tagSelection = [];
    let newTags = [];
    let edit = true;
    let dropDown = false;

    const setTagSelection = jest.fn(async (newTagsSelection) => tagSelection = newTagsSelection);
    const setNewTags = jest.fn(async (newTagSet) => newTags = newTagSet);
    const setEdit = jest.fn(async (bool) => edit = bool);
    const setDropDown = jest.fn(async (bool) => dropDown = bool);

    React.useEffect = jest.fn()
        .mockImplementation((effect, deps = null) => effect());
    
    React.useState = jest.fn()
        .mockReturnValueOnce([tagSelection, setTagSelection])
        .mockReturnValueOnce([newTags, setNewTags])
        .mockReturnValueOnce([edit, setEdit])
        .mockReturnValueOnce([dropDown, setDropDown]);

    const wrapper = mount(<TagList {...props} />);

    const nativeSyntheticEvent = {
        nativeEvent: {
            text: 'Artificial Intelligence',
        }
    }

    const textInput = wrapper
        .find('TextInput')
        .first();

    textInput.props().onEndEditing(nativeSyntheticEvent);
    wrapper.update();

    await new Promise(resolve => setImmediate(resolve));
    expect(newTags.length).toStrictEqual(1);
});

test('Save', async () => {

    let alertMessage = {};
    let visible = '';

    const props = {
        bookmarkId: '047ec796-2eca-417e-865a-e2e187b97689',
        currentTags: [], 
        setAlertMessage: jest.fn(async (newMessage) => alertMessage = newMessage),
        setVisible: jest.fn(async (bool) => visible = bool),
    };

    let tagSelection = [];
    let newTags = [];
    let edit = true;
    let dropDown = false;

    const setTagSelection = jest.fn(async (newTagsSelection) => tagSelection = newTagsSelection);
    const setNewTags = jest.fn(async (newTagSet) => newTags = newTagSet);
    const setEdit = jest.fn(async (bool) => edit = bool);
    const setDropDown = jest.fn(async (bool) => dropDown = bool);

    React.useEffect = jest.fn()
        .mockImplementation((effect, deps = null) => effect());
    
    React.useState = jest.fn()
        .mockReturnValueOnce([tagSelection, setTagSelection])
        .mockReturnValueOnce([newTags, setNewTags])
        .mockReturnValueOnce([edit, setEdit])
        .mockReturnValueOnce([dropDown, setDropDown]);

    updateBookmark.mockReturnValueOnce(new Promise((resolve, reject) => {
        resolve({
            status: 200,
            message: `Bookmark ${props.bookmarkId} Updated Successfully!`
        })
    }))

    const wrapper = mount(<TagList {...props} />);

    const saveIcon = wrapper
        .find('IconButton')
        .filterWhere(iconButton => iconButton.props().icon === 'content-save')
        .first();

    saveIcon.props().onPress();
    wrapper.update();

    await new Promise(resolve => setImmediate(resolve));
    expect(alertMessage.status).toStrictEqual('success');
})