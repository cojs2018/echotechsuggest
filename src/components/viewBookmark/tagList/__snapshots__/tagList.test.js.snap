// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagList renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "flex": 0,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
            "textAlign": "left",
          },
          undefined,
        ]
      }
    >
      Tags
    </Text>
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": undefined,
        }
      }
      accessible={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      <View>
        <Text
          accessibilityElementsHidden={true}
          allowFontScaling={false}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontSize": 24,
              },
              Array [
                Object {
                  "lineHeight": 24,
                  "transform": Array [
                    Object {
                      "scaleX": 1,
                    },
                  ],
                },
                Object {
                  "backgroundColor": "transparent",
                },
              ],
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          ó°£•
        </Text>
      </View>
    </View>
  </View>
  <View>
    <View>
      <View
        style={
          Object {
            "backgroundColor": "#ebebeb",
            "borderColor": "#ebebeb",
            "borderRadius": 16,
            "borderStyle": "solid",
            "borderWidth": 0.5,
            "elevation": 0,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": false,
              "selected": false,
            }
          }
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 16,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "paddingLeft": 4,
                  "position": "relative",
                },
                Object {
                  "paddingRight": 4,
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "lineHeight": 24,
                      "marginVertical": 4,
                      "minHeight": 24,
                      "textAlignVertical": "center",
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "marginLeft": 8,
                      "marginRight": 8,
                    },
                    undefined,
                  ],
                ]
              }
            >
              Artificial Precipitation
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#ebebeb",
            "borderColor": "#ebebeb",
            "borderRadius": 16,
            "borderStyle": "solid",
            "borderWidth": 0.5,
            "elevation": 0,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": false,
              "selected": false,
            }
          }
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 16,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "paddingLeft": 4,
                  "position": "relative",
                },
                Object {
                  "paddingRight": 4,
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "lineHeight": 24,
                      "marginVertical": 4,
                      "minHeight": 24,
                      "textAlignVertical": "center",
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "marginLeft": 8,
                      "marginRight": 8,
                    },
                    undefined,
                  ],
                ]
              }
            >
              Energy
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#ebebeb",
            "borderColor": "#ebebeb",
            "borderRadius": 16,
            "borderStyle": "solid",
            "borderWidth": 0.5,
            "elevation": 0,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": false,
              "selected": false,
            }
          }
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 16,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "paddingLeft": 4,
                  "position": "relative",
                },
                Object {
                  "paddingRight": 4,
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "lineHeight": 24,
                      "marginVertical": 4,
                      "minHeight": 24,
                      "textAlignVertical": "center",
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "marginLeft": 8,
                      "marginRight": 8,
                    },
                    undefined,
                  ],
                ]
              }
            >
              Quantum Computing
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#ebebeb",
            "borderColor": "#ebebeb",
            "borderRadius": 16,
            "borderStyle": "solid",
            "borderWidth": 0.5,
            "elevation": 0,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": false,
              "selected": false,
            }
          }
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 16,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "paddingLeft": 4,
                  "position": "relative",
                },
                Object {
                  "paddingRight": 4,
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "lineHeight": 24,
                      "marginVertical": 4,
                      "minHeight": 24,
                      "textAlignVertical": "center",
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "marginLeft": 8,
                      "marginRight": 8,
                    },
                    undefined,
                  ],
                ]
              }
            >
              Biotechnology
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
